From fcc7d51144872d4be76859b461d75f3a5b9f7668 Mon Sep 17 00:00:00 2001
From: Ihor Usyk <ihor_usyk@epam.com>
Date: Mon, 12 Jun 2023 16:45:27 +0200
Subject: [PATCH 08/12] sip_sv: rollback the service to support scmi

The SIP serviceis necessary because SCMI communication is based
at this service

Signed-off-by: Ihor Usyk <ihor_usyk@epam.com>
---
 plat/renesas/common/include/rcar_sip_svc.h | 33 ++++++++++++
 plat/renesas/rcar/platform.mk              |  3 +-
 plat/renesas/rcar/rcar_sip_svc.c           | 63 ++++++++++++++++++++++
 3 files changed, 98 insertions(+), 1 deletion(-)
 create mode 100644 plat/renesas/common/include/rcar_sip_svc.h
 create mode 100644 plat/renesas/rcar/rcar_sip_svc.c

diff --git a/plat/renesas/common/include/rcar_sip_svc.h b/plat/renesas/common/include/rcar_sip_svc.h
new file mode 100644
index 000000000..d9cd173d0
--- /dev/null
+++ b/plat/renesas/common/include/rcar_sip_svc.h
@@ -0,0 +1,33 @@
+/*
+ * Copyright (c) 2018-2020, Renesas Electronics Corporation. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#ifndef RCAR_SIP_SVC_H__
+#define RCAR_SIP_SVC_H__
+
+#include <lib/utils_def.h>
+
+
+/* General Service Queries */
+#define RCAR_SIP_SVC_CALL_COUNT		U(0x8200ff00)
+#define RCAR_SIP_SVC_UID		U(0x8200ff01)
+#define RCAR_SIP_SVC_VERSION		U(0x8200ff03)
+
+/* Function ID to get the Board type */
+#define RCAR_SIP_SVC_GET_BOARD_TYPE	U(0x82000001)
+
+/* Rcar SiP Service Calls version numbers */
+#define RCAR_SIP_SVC_VERSION_MAJOR	U(0x0)
+#define RCAR_SIP_SVC_VERSION_MINOR	U(0x2)
+
+/* Number of function IDs excluding general service queries */
+#define RCAR_SIP_SVC_FUNCTION_NUM	U(0x1)
+
+#define RCAR_SMC_RET_SUCCESS		0
+#define RCAR_SMC_RET_EFAILED		-2
+#define RCAR_SMC_RET_PMIC_DISABLE	-3
+
+
+#endif /* RCAR_SIP_SVC_H__ */
\ No newline at end of file
diff --git a/plat/renesas/rcar/platform.mk b/plat/renesas/rcar/platform.mk
index bebe3c108..7b3c20c2d 100644
--- a/plat/renesas/rcar/platform.mk
+++ b/plat/renesas/rcar/platform.mk
@@ -333,7 +333,8 @@ ifeq (${RCAR_GEN3_ULCB},1)
 BL31_SOURCES		+=	drivers/renesas/rcar/cpld/ulcb_cpld.c \
 				plat/renesas/rcar/rcar_scmi_power.c   \
 				plat/renesas/rcar/rcar_scmi_reset.c   \
-				plat/renesas/rcar/rcar_scmi_clocks.c
+				plat/renesas/rcar/rcar_scmi_clocks.c  \
+        plat/renesas/rcar/rcar_sip_svc.c
 endif
 
 # build the layout images for the bootrom and the necessary srecords
diff --git a/plat/renesas/rcar/rcar_sip_svc.c b/plat/renesas/rcar/rcar_sip_svc.c
new file mode 100644
index 000000000..1c9884065
--- /dev/null
+++ b/plat/renesas/rcar/rcar_sip_svc.c
@@ -0,0 +1,63 @@
+/*
+ * Copyright (c) 2016-2017, ARM Limited and Contributors. All rights reserved.
+ * Copyright (c) 2018-2021, Renesas Electronics Corporation. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+*/
+
+#include <common/runtime_svc.h>
+#include <tools_share/uuid.h>
+#include <common/debug.h>
+#include <rcar_sip_svc.h>
+#include "board.h"
+
+/*
+ * This function handles Rcar defined SiP Calls
+*/
+static uintptr_t rcar_sip_handler(unsigned int smc_fid,
+			u_register_t x1,
+			u_register_t x2,
+			u_register_t x3,
+			u_register_t x4,
+			void *cookie,
+			void *handle,
+			u_register_t flags)
+{
+#if PMIC_ROHM_BD9571
+	int32_t		ret;
+#endif	/* PMIC_ROHM_BD9571 */
+	/* Rcar SiP Service UUID */
+	DEFINE_SVC_UUID2(rcar_sip_svc_uid,
+			0x120f8108U, 0x8152U, 0x4154U, 0xa2U, 0xafU,
+			0x7dU, 0x8aU, 0x7eU, 0x6dU, 0x90U, 0xeeU);
+
+	switch (smc_fid) {
+	case RCAR_SIP_SVC_CALL_COUNT:
+		/* Return the number of function IDs */
+		SMC_RET1(handle, RCAR_SIP_SVC_FUNCTION_NUM);
+
+	case RCAR_SIP_SVC_UID:
+		/* Return UID to the caller */
+		SMC_UUID_RET(handle, rcar_sip_svc_uid);
+
+	case RCAR_SIP_SVC_VERSION:
+		/* Return the version of current implementation */
+		SMC_RET2(handle, RCAR_SIP_SVC_VERSION_MAJOR,
+			RCAR_SIP_SVC_VERSION_MINOR);
+
+	default:
+		WARN("Unimplemented Rcar SiP Service Call: 0x%x\n", smc_fid);
+		SMC_RET1(handle, SMC_UNK);
+	}
+
+}
+
+/* Define a runtime service descriptor for fast SMC calls */
+DECLARE_RT_SVC(
+	rcar_sip_svc,
+	(uint8_t)OEN_SIP_START,
+	(uint8_t)OEN_SIP_END,
+	(uint8_t)SMC_TYPE_FAST,
+	NULL,
+	rcar_sip_handler
+);
-- 
2.25.1

